service/history/statemachine_environment.go

```diff
@@ -313,13 +304,11 @@ func (e *stateMachineEnvironment) validateStateMachineRefWithoutTransitionHistor
 		return fmt.Errorf("%w: state machine ref initial failover version mismatch", consts.ErrStaleReference)
 	}
 
-	// Only check for strict equality if the ref has non zero MachineTransitionCount, which marks the task as non-concurrent.
-	if ref.StateMachineRef.MachineTransitionCount != 0 &&
-		node.InternalRepr().GetTransitionCount() != ref.StateMachineRef.MachineTransitionCount {
-		return fmt.Errorf("%w: state machine transitions != ref transitions", consts.ErrStaleReference)
+	// This is only expected to be set on tasks for now.
+	if ref.Validate == nil {
+		return nil
 	}
-
-	return nil
+	return ref.Validate(ref.StateMachineRef, node)
```

Comment:
we need this in `validateStateMachineRef` as well I think.

Reply:
Good catch!